### ISSUE TITLE
A need for Addons to have own sound channel

### ISSUE THREAD

**Issue by MysticalOS** [Stanzilla/WoWUIBugs#79]
With patch 7.3.5 and beyond, there were sound changes that affected addons, as you know. Can get full backstory of stuff here: https://www.patreon.com/posts/addon-sound-work-16575915 (public post, just where I happened to put it at the time)

It's been over 3 years since patch launched and the changes in 7.3.5 have NOT been reverted (at least not in full, I know they made some tweaks but they ultimately never really solved problem with addon sounds that patch introducedd. Might never be. Maybe it's intended for SFX to now have higher priority than MASTER (when it previous did not). If that's case, addons will probably start pushing their sounds to DIALOG more and more. Other addons may force 64 or even 128 sound channels without users consent like Bigwigs and DBM currently do by forcing 64 channels on users if they are set lower than 64.

However, instead of addons hijacking ANY audio channel or hijacking user settings, it's high time that addons got their own dedicated channel that has high priority. Currently we have DIALOG > SFX > MASTER (with MUSIC and AMBIENT in there somewhere, kinda hard to test where they fit in priority though. Imagine if we had DIALOG > ADDON > SFX > MASTER.
You'll probably still always want dialog first in the super and unlikely off chance somehow addon sounds exceed channels all on their own without help of SFX and you won't want to ruin users experiencing story/dialog because of it. That'd be fine though cause that'd probably never happen. However, with "Addon" channel in front of SFX that's all we'd never need. It's basically what we had when it was MASTER > SFX pre 7.3.5.

So you might wonder, "well if we just flip master and sfx back around, is that enough". Yes and no. To be honest, addons have used master because we had to, not so much by choice. However we get constant user complaints "why my addon sounds so much higher/lower than x? "How do I adjust addon sounds withoutout affecting other sounds?", etc. We end up saying "you can't. no matter what channel addon uses, it's always shared volume with something else".

That's another thing I'd like to finally see fixed. With an "addon" channel we also get an ADDON ONLY volume slider to finally provide users the control they need that does not tie the volume of their addons to some other volume slider like DIALOG, SFX, or MASTER. They can control the volume of their addons in a way that's never been possible before and that'd be a fantastic addition to game.

If things continue as is, with no change, addons will continue to highjack user settings to mitigate the problems with their experience and continue to have volume problems with other sounds/events since we're force to share. The experience is always diminished in some way because of this. You end up with users that turn master down, but then have to turn everything else up a little, but then when they get a skype message or something their eardrums are blown out because they don't realize how high they've turned their speaker volume because obviously adjusting master slider affects sound output of WoW globally. It's more complicated than it needs to be just for users to volume tune their addons.

TL/DR
We need an "ADDON" sound channel and volume slider. I'd like to see a patch where master/sfx/dialog sliders  become a slider users use for just those things, not their addons. I want to be able to direct users of my mod to clean solution that just involves them adjusting an addon slider to control addon volume for an addon channel that will never be victim of sound ducking.

**Comment by LenweSaralonde** [comment]
Another possible approach would be to be able to adjust the volume of sounds played by add-ons:
* Add a new `volume` parameter to `PlaySoundFile()`, `PlaySound()` and why not `PlayMusic()`.
* Add new functions such as `SetSoundVolume(handle, volume)` and  `volume = GetSoundVolume(handle)` to adjust the volume of a sound being played.

`volume` being a number value between 1 (default, 100%) and 0.

This would allow to play some alerts louder/quieter than others.

Some sounds being played through  `SFX` have a different behavior, such as being affected by environmental reverb etc. so "add-ons sound channels" should be independent to the the Master, SFX and Music channels.

**Comment by MysticalOS** [comment]
That could work in part, but it's also a new complication to deal with for both authors and users.

Not to mention it's not global which i feel is more sensicle to end users that aren't power users. Most aren't gonna go "ok I need every single addon to be diff volume" for most part users just want addons to be at one volume and not have it affect volume of other stuff. A dedicated channel solves that. Same with being able to enable/disable addon sounds independantly without also tying it to another channel it has to share with. On our end we simply move our calls to the established "ADDON" channel and call it a day. Lastly it's just least amount of work for blizzard. They aren't likely to invest too much time into a complex system

**Comment by LenweSaralonde** [comment]
> Other addons may force 64 or even 128 sound channels without users consent like Bigwigs and DBM currently do by forcing 64 channels on users if they are set lower than 64.

I've made some tests using [Musician](https://github.com/LenweSaralonde/Musician) I'm working on to determine how many sounds the UI can play simultaneously using the macro below (this will play a bagpipe drone sound until PlaySoundFile() returns false, meaning the limit has been reached):
```lua
/run local p, tf = 0; tf = function() if PlaySoundFile('Interface\\AddOns\\Musician\\instruments\\bagpipe\\C3.ogg', 'SFX') then p = p + 1; C_Timer.After(1/30, tf); else print("Max polyphony", p) end end; tf();
```
* Using `SFX`, the max amount of sounds I could play simultaneously is **15**, regardless of the number of sound channels
* Using `Master`, this amount is **30** for 48 and 64 sound channels but seems unlimited while using 24 or any custom value such as 128.

This was tested on Retail 9.0.5.38556

It's quite interesting to see how the `Sound_NumChannels` setting affects UI polyphony. `Master` is a good choice for add-ons that need to play a lot simultaneous sounds but it can go insanely loud without any way to the end user to adjust the volume.

**Comment by MysticalOS** [comment]
That's interesting info. I know the issue wasn't so much addons needing to play a lot of sounds at once though, the issue was that when users have SFX on, the number of non addon sounds going on was choking out addon sounds from playing at all (even if there weren't a lot of them).

Plus the issues with configuring addon sound volume still being tied to another sound channels volume still.

But you're saying with your findings that setting it to 64 actualy caps it at 24 but setting it to 24 made it unlimitted? 🤦‍♀️

**Comment by LenweSaralonde** [comment]
> your findings that setting it to 64 actualy caps it at 24 but setting it to 24 made it unlimitted?
No, It's not unimited, it actually just causes the whole sound engine to glitch.

I remade the experiment using a harp sound (6-second sound starting with a peak and decreasing) and audio recording to find out exactly what's going on:
```lua
/run local p, tf = 0; tf = function() if PlaySoundFile('Interface\\AddOns\\Musician\\instruments\\harp\\C4.ogg', 'Master') then p = p + 1; C_Timer.After(0, tf); else print("Max polyphony", p) end end; tf(); 
```
I run the macro after a fresh game start 

* Using 48, 64 and 128 Sound_NumChannels => The macro stops at Max polyphony 30, all sounds play properly properly heard and other game sounds play correctly
* Using 24 Sound_NumChannels => The macro keeps running but sound gets stuttering, all game sounds stop working. Sounds like there is a kind of circular

My conclusion is Sound_NumChannels DOES NOT affect the cap for simultaneous UI sounds (15 for SFX, 30 for Master) on PlaySound / PlaySoundfile but if the UI attempts to play more sounds than set in Sound_NumChannels, other currently playing sounds will stop playing. It would make more sense to have this cap to be a percentage of Sound_NumChannels. Anyway, this is another problem than the originally exposed one about channel volume but would be worth its own ticket.

**Comment by LenweSaralonde** [comment]
I made another funny discovery.

Each channel has its own cap:
* Master: 30
* SFX: 15
* Dialog: 20
* Music: 20 (when enabled)

But each cap is independent to each other. PlaySoundFile() can play perfectly 30 simultaneous files in the Master channel + 15 in SFX + 20 in Dialog as long as the total amount of playing files remains under Sound_NumChannels. 

The macro below stacks the D.I.S.C.O. ball opening sound in Master, SFX then Dialog until all available slots are full. With Sound_NumChannels set to 128, Max polyphony 65 is reached . (Beware, it's loud):
```lua
/run local ch, i, p, tf = {'Master', 'SFX', 'Dialog'}, 1, 0; tf = function() if PlaySoundFile(567381, ch[i]) then p = p + 1; C_Timer.After(0, tf); elseif ch[i+1] then i = i + 1; C_Timer.After(0, tf); else print("Max polyphony", p) end end; tf();
```

**Comment by MysticalOS** [comment]
So yeah setting channels to 64 kinda makes sure the caps can actually be filled, but doesn't explain sound ducking on master channel when SFX fills up. MANY user reports state that addon sounds get muted/ducked when they are reaching SFX cap, even though we're in master, so you are sure each cap is independant? if masters 30 isn't consumed by SFX, why do addons playing in master lose sounds unless user turns off SFX?

I feel like there was a cap of how many can be played at same time universally, maybe on top of these where even at 128 channels, addon sounds can still get muted if too much going on. I did find in these events a priority seemed to kick in that was something like Dialog > SFX > Master so master sounds were first to get squashed.

It's all been a mess since 7.x patch that changed it.

**Comment by LenweSaralonde** [comment]
Each cap is independent for UI sounds played using PlaySoundFile() in the sense that PlaySoundFile() will return a valid sound handle depending on the channel cap and usage. But if the global channel cap set by Sound_NumChannels is exceeded, some sounds won't play anyway (or other sounds will get muted/ducked), even if a valid sound handle is returned by PlaySoundFile().
